### PR TITLE
Add l10n_ro_partner_create_by_vat module.

### PR DESCRIPTION
**Pull request by feketemihai**
Add module to create partners based on the VAT numbers, The datas are fetched from multiple sources, see README file.

### CODE REVIEW

File: partner_create_by_vat/views/res_partner_view.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+            <field name="model">res.partner</field>
+            <field name="inherit_id" ref="base_vat.view_partner_form"/>
+            <field name="arch" type="xml">
+                <xpath expr="//sheet" position="before">
+           		    <header>
+ 				        <button name="button_get_partner_data" attrs="{'invisible': [('name', '&lt;&gt;', ' '),('vat', '=', False),('is_company', '=', False )]}" colspan="2" string="Get Partner Data" context="{'active_id': active_id}" type="object" icon="gtk-execute"/>
+                	</header>  
+           	    </xpath>
+                <label for="vat" position="replace"/>
+                <div name="vat_info" position="replace">
```
Comment (StefanRijnhart) [inline partner_create_by_vat/views/res_partner_view.xml]: You may be replacing too much here, potentially losing compatibility with other vat related modules. Is it just the 'Check VAT' button you want to remove?

File: partner_create_by_vat/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                            if state:
+                                state = state[0].id
+                        self.write({
+                            'name': res['name'].upper(),
+                            'nrc': res['registration_id'] and
+                            res['registration_id'].upper() or '',
+                            'street': res['address'].title(),
+                            'city': res['city'].title(),
+                            'phone': res['phone'] and res['phone'] or '',
+                            'fax': res['fax'] and res['fax'] or '',
```
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: You can just say 

```
'fax': res['fax'] or '',
```

Not using res.get('fax') does imply that 'fax' is always present in the dictionary, but may be None or False or so.

File: partner_create_by_vat/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                        jud = result['Judetul:'].title() or ''
+                        if jud.lower().startswith('municip'):
+                            jud = ' '.join(jud.split(' ')[1:])
+                        if jud != '':
+                            state = self.env['res.country.state'].search(
+                                [('name', 'ilike', jud)])
+                            if state:
+                                state = state[0].id
+                    if 'Telefon:' in result.keys():
+                        tel = result['Telefon:'].replace('.', '') or ''
```
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: The " or ''" bit is void here. If you can call _replace_ on something, it is most likely already a string. It may be an empty string, in which case you assign an empty string as well.

File: partner_create_by_vat/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                        'name': name or '',
+                        'nrc': nrc or '',
+                        'street': adresa or '',
+                        'phone': tel or '',
+                        'fax': fax or '',
+                        'zip': zip1 or '',
+                        'vat_subjected': vat_s or False,
+                        'state_id': state,
+                    })
+                except:
```
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: It is usually bad practice to catch any exception instead of a specific one, especially on such a large block of code.

File: partner_create_by_vat/views/res_partner_view.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+                     <field name="vat" on_change="base_vat.vat_change(vat)" placeholder="e.g. BE0477472701" class="oe_inline"/>
```
Comment (StefanRijnhart) [inline partner_create_by_vat/views/res_partner_view.xml]: Does that work, appending a module name to a method as in 'base_vat.vat_change(vat)'?

File: partner_create_by_vat/models/res_partner.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
+# -*- coding: utf-8 -*-
+# ©  2015 Forest and Biomass Services Romania
+# See README.rst file on addons root folder for license details
+
+import requests
+
+from string import maketrans
+from stdnum.eu.vat import check_vies
```
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: Please import check_vies where you use it. That way, people who do not actually have the module installed do not have to have the stdnum module in their setup.

File: partner_create_by_vat/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+from lxml import html
+
+from openerp import models, fields, api, _
+from openerp.exceptions import except_orm
+
+CEDILLATRANS = maketrans(u'\u015f\u0163\u015e\u0162'.encode(
+    'utf8'), u'\u0219\u021b\u0218\u021a'.encode('utf8'))
+
+
+def getMfinante(cod):
```
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: Make this a  method so to be inheritable?

File: partner_create_by_vat/models/res_partner.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    name = fields.Char('Name', required=True, select=True, default=' ')
+
+    @api.one
+    def button_get_partner_data(self):
+        def _check_vat_ro(vat):
```
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: This inline function has a vat argument, but refers to part.name instead. Apparently it works, but I think you meant something more elegant.

**Comment by StefanRijnhart** [comment]
Nice module! Would you consider renaming to l10n_ro_partner_create_by_vat or something? Then we may generalize this module for more general use and make this module an implemenation of that module, adding Romanian specifics.

**Comment by feketemihai** [comment]
@StefanRijnhart Before fixing the comments, do you think it's usefull to split the module in 2, to have the base on creating partners based the vies datas (probably submitted to partner-contact repo), and to move the Romanian creation in l10n_ro_partner_create_by_vat...and in this case, we can inherit the vies module creation...

File: partner_create_by_vat/models/res_partner.py (hunk 8 above)
Comment (feketemihai) [inline partner_create_by_vat/models/res_partner.py]: This is for easier creation, basically when you push the create button on a partner, your first field is the name, (modified default to ' '), not to raise error if you only complete the vat field and push the button.
So, when creating a partner, you can put the vat_number in the name field, and if doesn't have the vat field completed, it checks if you put the vat number into the name field...

File: partner_create_by_vat/models/res_partner.py (hunk 6 above)
Comment (feketemihai) [inline partner_create_by_vat/models/res_partner.py]: You still need the stdnum, because if it's not a romanian company, it will use the vies service, and by default vatnumber from python doesn't retrieve the fields: name, adress from vies like stdnum does...

File: partner_create_by_vat/views/res_partner_view.xml (hunk 1 above)
Comment (feketemihai) [inline partner_create_by_vat/views/res_partner_view.xml]: I don't quite like the base_vat div, since the label is on the div, so you have vat, the button and vat_subjected field in the same labelling, and you really don't know what the tickbox (vat_subjected) represents...

File: partner_create_by_vat/views/res_partner_view.xml (hunk 5 above)
Comment (feketemihai) [inline partner_create_by_vat/views/res_partner_view.xml]: i don't have a error since using it, even if the moethod it's not doing quite anything...

File: partner_create_by_vat/views/res_partner_view.xml (hunk 1 above)
Comment (StefanRijnhart) [inline partner_create_by_vat/views/res_partner_view.xml]: I'd say that this change is out of scope for this module then.

Comment (feketemihai) [inline partner_create_by_vat/views/res_partner_view.xml]: @StefanRijnhart I agree and will change it...can you check lower comment about splitting the module in 2?

File: partner_create_by_vat/models/res_partner.py (hunk 6 above)
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: Yes, but the problem is another: the web client loads every module in the addons paths if it has a static directory, even if they're not installed. When whole repository branches are added to the addons path, which is not uncommon, you don't want to be installing all the python imports for modules you are not using.

You have the external dependency correctly noted in the manifest file, so you can safely move the import to the method where it does no harm anymore. This is a common OCA convention.

File: partner_create_by_vat/models/res_partner.py (hunk 8 above)
Comment (StefanRijnhart) [inline partner_create_by_vat/models/res_partner.py]: Yes, but I was just talking about the vat argument in the signature of _check_vat_ro(vat). It is unused in the rest of this inline function so I was wondering if you meant to refer to something else or if it can be left out entirely.

File: partner_create_by_vat/views/res_partner_view.xml (hunk 5 above)
Comment (StefanRijnhart) [inline partner_create_by_vat/views/res_partner_view.xml]: Maybe the method is not called anymore, because the function reference is invalid now?

**Comment by StefanRijnhart** [comment]
About the splitting, that would be the best!

**Comment by feketemihai** [comment]
@StefanRijnhart I left the VIES creation to partner_create_by_vat, and submitted it to partner-contact repo. See https://github.com/OCA/partner-contact/pull/216.
I will do the l10n_ro_partner_create_by_vat today.

**Comment by feketemihai** [comment]
@StefanRijnhart Can i have a review on the updates...

**Comment by feketemihai** [comment]
Travis and runbot needs https://github.com/OCA/partner-contact/pull/216 to be merged, only dependency error of module partner_create_by_vat.

**Comment by StefanRijnhart** [comment]
Thanks for all your work :+1:
